doc/tutorial/tutorial.md

```diff
@@ -1,40 +1,650 @@
+    $"Host: IP=${h.ip}, MAC=${h.mac}"
+}
+```
+
+DDlog will automatically invoke the user-defined string conversion functions for `ip_addr_t` and
+`mac_addr_t` types to format `ip` and `mac` fields respectively.
+
+The following DDlog program tests the above types and functions.  It maps values of type `nethost_t`
+in the input relation to string representation and stores them in the derived relation:
+
+```
+input relation NetHost(id: bigint, h: nethost_t)
+relation NetHostString(id: bigint, s: string)
+
+NetHostString(id, $"${h}") :- NetHost(id, h).
+```
+Try feeding the following `.dat` file to this program:
+
+```
+start;
+
+insert NetHost(1, NHost{IPAddr{0xaabbccdd}, MACAddr{0x112233445566}}),
+insert NetHost(2, NHost{IPAddr{0xa0b0c0d0}, MACAddr{0x102030405060}});
+
+commit;
+
+dump NetHostString;
+```
+
+You should obtain the following output:
+
+```
+NetHostString{1,"Host: IP=170.187.204.221, MAC=11:22:33:44:55:66"}
+NetHostString{2,"Host: IP=160.176.192.208, MAC=10:20:30:40:50:60"}
+```
+
+### Bit vectors
+
+The design of DDlog was strongly influenced by applications from the domain of networking, where
+programmers frequently deal with bit vectors of non-standard sizes.  This motivated DDlog's
+bit-vector type.  We have encountered several instances of this type, e.g., `typedef UUID=bit<128>`.
+In addition to standard arithmetic and bitwise operators over bit vectors (see [language
+reference](../language_reference/language_reference.md#expressions)), DDlog supports bit vector
+slicing and concatenation.  Slicing is expressed using the `x[N:M]` syntax examplified in the
+previous section.  Bit vector concatenation is implemented by the `++` operator (which is also
+overloaded for strings).  Concatenation can be applied to bit vectors of arbitrary width, producing
+a bit vector whose width is the sum of the widths of its arguments.  The following example
+illustrates both operators:
+
+```
+// Form IP address from bytes using bit vector concatenation
+function ip_from_bytes(b3: bit<8>, b2: bit<8>, b1: bit<8>, b0: bit<8>)
+    : ip_addr_t =
+{
+    IPAddr{.addr = b3 ++ b2 ++ b1 ++ b0}
+}
+
+// Check for multicast IP address using bit slicing
+function is_multicast_addr(ip: ip_addr_t): bool = ip.addr[31:28] == 14
+
+input relation Bytes(b3: bit<8>, b2: bit<8>, b1: bit<8>, b0: bit<8>)
+
+// convert bytes to IP addresses
+relation Address(addr: ip_addr_t)
+Address(ip_from_bytes(b3,b2,b1,b0)) :- Bytes(b3,b2,b1,b0).
+
+// filter multicast IP addresses
+relation MCastAddress(addr: ip_addr_t)
+MCastAddress(a) :- Address(a), is_multicast_addr(a).
+```
+
+DDlog offers two ways to write bit vector literals. First, they can be written as decimal numbers.
+DDlog performs *type inference* to determine whether the number should be interpreted as an unbounded
+mathematical integer or a bit vector.  In the latter case it also infers its bit width.  For
+example, the literal `125` is interpreted as a mathematical integer in `var x: int = 125` and as an
```

Comment:
I think now it's bigint.
I would rather have a separate paragraph to introduce bigint.